### PR TITLE
Multisig wallet

### PR DESCRIPTION
**Pull request by puzzle-rusher**

### CODE REVIEW

File: multisig-wallet/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,508 @@
...
+        let confirmation = self
+            .confirmations
+            .entry(*transaction_id)
+            .or_insert(BTreeMap::new())
+            .entry(msg::source())
+            .or_insert(true);
+
+        *confirmation = true;
+
+        self.execute_transaction(transaction_id, None::<fn()>);
```
Comment (shamilsan) [inline multisig-wallet/src/lib.rs:218]: Why not just `None`?
```suggestion
        self.execute_transaction(transaction_id, None);
```

File: multisig-wallet/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,508 @@
...
+    payload: Vec<u8>,
+    value: u128,
+    executed: bool,
+}
+
+#[derive(Default)]
+pub struct MultisigWallet {
+    pub transactions: BTreeMap<U256, Transaction>,
+    pub confirmations: BTreeMap<U256, BTreeMap<ActorId, bool>>,
+    pub is_owner: BTreeMap<ActorId, bool>,
```
Comment (shamilsan) [inline multisig-wallet/src/lib.rs:27]: Isn't better to use `BTreeSet` here?

File: multisig-wallet/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,508 @@
...
+    pub owners: Vec<ActorId>,
```
Comment (shamilsan) [inline multisig-wallet/src/lib.rs:28]: May we combine `owners` and `is_owner` into one collection?

File: multisig-wallet/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,508 @@
...
+    /// transactionId Transaction ID.
+    fn revoke_confirmation(&mut self, transaction_id: &U256) {
+        self.validate_owner_exists(&msg::source());
+        self.validate_confirmed(transaction_id, &msg::source());
+        self.validate_not_executed(transaction_id);
+
+        let confirmation = self
+            .confirmations
+            .entry(*transaction_id)
+            .or_insert(BTreeMap::new())
```
Comment (shamilsan) [inline multisig-wallet/src/lib.rs:244]: A little bit shorter:
```suggestion
            .or_default()
```

File: multisig-wallet/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,508 @@
...
+pub struct Transaction {
+    destination: ActorId,
+    payload: Vec<u8>,
+    value: u128,
+    executed: bool,
+}
+
+#[derive(Default)]
+pub struct MultisigWallet {
+    pub transactions: BTreeMap<U256, Transaction>,
```
Comment (shamilsan) [inline multisig-wallet/src/lib.rs:25]: I'd add a type alias for transaction id for readability.
```rust
type TransactionId = U256;
```

File: multisig-wallet/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,508 @@
...
+                self.confirmations
+                    .get(transaction_id)
+                    .and_then(|confirmations| confirmations.get(owner))
+                    .copied()
+                    .unwrap_or(false)
+            })
+            .filter(|confirm| *confirm)
+            .count()
+            .try_into()
+            .unwrap()
```
Comment (shamilsan) [inline multisig-wallet/src/lib.rs:352]: Also, you can use `as` here.

File: multisig-wallet/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,508 @@
...
+    }
+
+    /// Returns total number of transactions after filers are applied.
+    /// pending Include pending transactions.
+    /// executed Include executed transactions.
```
Comment (shamilsan) [inline multisig-wallet/src/lib.rs:357]:
```suggestion
    /// * `pending` Include pending transactions.
    /// * `executed` Include executed transactions.
```

File: multisig-wallet/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,508 @@
...
+        input : MWAction,
+        output : MWEvent,
+    state:
+        input: State,
+        output: StateReply,
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn init() {
+    let config: MWInitConfig = msg::load().expect("Unable to decode InitDao");
```
Comment (shamilsan) [inline multisig-wallet/src/lib.rs:420]: Dao?

File: multisig-wallet/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,508 @@
...
+    }
+}
+
+gstd::metadata! {
+    title: "MultisigWallet",
+    init:
+        input : MWInitConfig,
+    handle:
+        input : MWAction,
+        output : MWEvent,
```
Comment (shamilsan) [inline multisig-wallet/src/lib.rs:412]: Spaces.
```suggestion
        input: MWInitConfig,
    handle:
        input: MWAction,
        output: MWEvent,
```

File: multisig-wallet/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,508 @@
...
+    WALLET = Some(wallet);
+}
+
+#[gstd::async_main]
+async unsafe fn main() {
+    let action = match msg::load::<MWAction>() {
+        Ok(action) => action,
+        Err(_) => {
+            let bytes: Vec<u8> = msg::load_bytes();
+            MWAction::decode(&mut bytes.as_slice()).expect("Could not load Action")
```
Comment (shamilsan) [inline multisig-wallet/src/lib.rs:448]: Isn't it the same as `msg::load::<MWAction>()`?

**Review by shamilsan — COMMENTED** [review]
Generally, I'd consider using `BTreeSet<T>` instead of `BTreeMap<T, bool>` where it is possible.

File: multisig-wallet/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,485 @@
...
+        validate_requirement(
+            next_owners_count,
+            min(next_owners_count as u64, self.required),
+        );
+
+        self.is_owner.remove(owner);
+        self.owners.retain(|&x| x != *owner);
+
+        if (self.owners.len() as u64) < self.required {
+            self.change_requirement(self.owners.len().try_into().unwrap());
```
Comment (shamilsan) [inline multisig-wallet/src/lib.rs:145]:
```suggestion
            self.change_requirement(self.owners.len() as u64);
```

**Review by shamilsan — COMMENTED** [review]
Need to resolve fixed conversations and to comment that are not fixed.

**Review by LouiseMedova — CHANGES_REQUESTED** [review]
Looks good but tests can be written mush shorter

File: multisig-wallet/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,480 @@
...
+        }
+    }
+
+    fn validate_not_executed(&self, transaction_id: &TransactionId) {
+        if matches!(self.transactions.get(transaction_id), Some(t) if t.executed) {
+            panic!("Transaction has been already executed")
+        }
+    }
+
+    fn has_owner(&self, owner: &ActorId) -> bool {
```
[the remark this answers is not in the document]
Comment (puzzle-rusher) [reply multisig-wallet/src/lib.rs:109]: I use it like this `self.has_owner(owner)` where self is wallet in this context, so I thought it was more expressive.

File: multisig-wallet/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,480 @@
...
+
+        count >= self.required
+    }
+
+    /// Adds a new transaction to the transaction mapping, if transaction does not exist yet.
+    /// `destination` Transaction target address.
+    /// `value` Transaction ether value.
+    /// `data` Transaction data payload.
+    /// Returns transaction ID.
+    fn add_transaction(
```
Comment (LouiseMedova) [inline multisig-wallet/src/lib.rs:295]: Need to checks that transaction value is equal to attached value

**Review by shamilsan — APPROVED** [review]
You'll have clippy warnings after merging master, please fix them.

**Comment by shamilsan** [comment]
Closing in favor https://github.com/gear-dapps/multisig-wallet/pull/1

**Pull request closed without merging**